feat: add support for declaring search indexes

noseworthy opened the pull request:
MongoDB and `mongoose` now support adding search indexes to collections. Introduce a new decorator `@searchIndex()` that, similar to `@index`, adds a search index to a collection.

Search index support was added to `mongoose` in version `8.1.0`, however the type definition for `Schema.searchIndex()` didn't exist until `8.2.3`. Therefore in order to support this feature, update the `mongoose` dependencies to `~8.2.3`.

## Related Issues

- N/A

noseworthy commented:
@hasezoey : Sorry, the typings sort of got out of hand here. I can totally remove all that if you'd like. I know it's probably not `typegoose`'s responsibility. The typings in the mongo driver were so lacklustre though and when developing this I had to build up my own mental map of what was going on.

Anyway, It's pretty easy to simplify if you'd like. I can try and open a PR into the mongo driver repo and see if they'll take them on haha.

Also, wasn't sure how to handle the requirement on `mongoose` `8.2.3`. Let me know if I should've done something different there.

Finally, thanks so much for your work on `typegoose` and `mongoose`. These libraries are fantastic 😄

hasezoey commented on docs/api/decorators/searchIndexes.md:
````diff
@@ -0,0 +1,75 @@
+**Typings:**
+
+```ts
+function searchIndex<T extends BeAnObject = BeAnObject>(description: SearchIndexDescription): ClassDecorator
````
```suggestion
function searchIndex(description: SearchIndexDescription): ClassDecorator
```

`T` is not used anywhere and the code version also does not have it.

even with this change, it does not really match the code definition

hasezoey commented on docs/api/decorators/searchIndexes.md:
```diff
@@ -0,0 +1,75 @@
+
+| Name                                                               |                                                         Type                                                         | Description                                                    |
+|:-------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------:|:---------------------------------------------------------------|
+| `description` <span class="badge badge--secondary">Required</span> | [`SearchIndexDescription`](https://mongodb.github.io/node-mongodb-native/6.5/interfaces/SearchIndexDescription.html) | Description of the search index, including name and definition |
+
+`@searchIndex` is used to set search indices on the schema, this decorator acts like [`schema.searchIndex()`](https://mongoosejs.com/docs/api/schema.html#Schema.prototype.searchIndex()).
+
+:::note
+Because creating a search index can be a very heavy operation, automatic creation of search indices is disabled by
+default. To enable automatic creation of search indices, the [`autoSearchIndex`](https://mongoosejs.com/docs/guide.html#autoSearchIndex) option must be set to `true` in the
+schema options using the [`@modelOptions`](https://typegoose.github.io/typegoose/docs/api/decorators/model-options) decorator
+:::
```
the atlas only note should be moved up here with either a `note`, `tip` or `caution` admonition, [test page](https://typegoose.github.io/typegoose/docs/test/#callouts--admonitions)

hasezoey commented on docs/api/decorators/searchIndexes.md:
```diff
@@ -0,0 +1,75 @@
+
+## Extra
+
+Search indices are only supported in `M10` Mongo Atlas clusters running MongoDB 6.0+ or 7.0+. Full documentation
+can be found [here](https://www.mongodb.com/docs/atlas/atlas-search/manage-indexes/)
```
```suggestion
Search indices are only supported in `M10` Mongo Atlas clusters running MongoDB 6.0+ or 7.0+. Full documentation
can be found [here](https://www.mongodb.com/docs/atlas/atlas-search/manage-indexes/).
```

hasezoey reviewed (commented):
Thanks for the work, there are some small things i have noticed and put some comments up, aside from that there is also:
- the new documentation page has not been added to the sidebars yet
- i would prefer if those extra types could be upstreamed to mongoose (or even mongodb?)
- some places have been missed for the changed mongoose version

i will put up a update on the beta branch, please target that as a base

hasezoey commented:
i will not be able to upgrade mongoose until https://github.com/Automattic/mongoose/pull/14463 is merged and released (the same cause on why the test CI failed here)

noseworthy commented:
Thanks for the review @hasezoey ! I think I've addressed all the comments. I see that the `beta` branch is still a ways behind `master`, so once that updates I'll rebase on top of that.

I've created [NODE-6044](https://jira.mongodb.org/browse/NODE-6044) in the `mongodb` jira and an associated [PR](https://github.com/mongodb/node-mongodb-native/pull/4052). We'll see how the mongo team feels about that. It'd be great to have more detailed types other than just `bson.Document` for the search index definition 😅.

I've opened another [PR](https://github.com/Automattic/mongoose/pull/14464) in `mongoose` to re-export the `SearchIndexDescription` type from `mongodb` so that all three projects can stay in sync. Since I don't currently have access to this type, I've created a copy for the time being, but if that PR gets merged then I'll just remove the copy.

Thanks again for your patience with this and your review. I apologize in advance if there's any more dumb issues or if I've missed something. Still getting my feet under me in this repo 😅!

hasezoey commented:
> I think I've addressed all the comments.

from what i can see, yes

> I see that the beta branch is still a ways behind master, so once that updates I'll rebase on top of that.

i dont want to push broken test commits, so i have not pushed the mongoose 8.2.3 update there and wait until https://github.com/Automattic/mongoose/pull/14463 is resolved.
for the same reason this PR will likely not get merged for now.

hasezoey commented:
slight update, https://github.com/Automattic/mongoose/pull/14463 has been merged, now just waiting until 8.2.4 is released

hasezoey commented:
the mongoose version 8.2.4 has been released and the beta branch has been updated (and also released as [v12.3.0-beta.1](https://github.com/typegoose/typegoose/releases/tag/v12.3.0-beta.1)), please rebase onto that

noseworthy commented:
All done, @hasezoey! Thanks again for helping review this change. If there's anything else you'd like done, just let me know.
